File: sdk/python/examples/langchain_integration.py
"""LangChain semantic caching example.

Demonstrates using Ferrite as a transparent semantic cache for LangChain
LLM calls, reducing API costs for semantically similar queries.

Prerequisites:
    pip install ferrite-ai[langchain] langchain-openai
    export OPENAI_API_KEY="sk-..."

    # Start Ferrite server
    cargo run --release
"""

import time

from langchain.globals import set_llm_cache
from langchain_openai import ChatOpenAI

from ferrite_ai.langchain import FerriteSemanticCache


def main():
    # Configure Ferrite as the LangChain cache
    set_llm_cache(FerriteSemanticCache(
        host="127.0.0.1",
        port=6380,
        similarity_threshold=0.85,
        ttl=3600,
        namespace="lc_demo",
    ))

    llm = ChatOpenAI(model="gpt-4o-mini", temperature=0)
    print("LangChain + Ferrite Semantic Cache\n")

    queries = [
        ("What is the capital of France?", "Cache MISS (first call)"),
        ("France's capital city?", "Cache HIT (semantically similar)"),
        ("Explain quantum entanglement", "Cache MISS (different topic)"),
        ("What is quantum entanglement?", "Cache HIT (semantically similar)"),
    ]

    for query, expected in queries:
        t0 = time.time()
        response = llm.invoke(query)
        latency = (time.time() - t0) * 1000

        print(f"Query:    '{query}'")
        print(f"Response: {response.content[:100]}...")
        print(f"Latency:  {latency:.1f} ms — {expected}")
        print()


if __name__ == "__main__":
    main()

File: sdk/typescript/examples/langchain-integration.ts
/**
 * LangChain.js semantic caching example.
 *
 * Demonstrates using Ferrite as a transparent semantic cache for LangChain
 * LLM calls, reducing API costs for semantically similar queries.
 *
 * Prerequisites:
 *   npm install @ferrite/ai @langchain/core @langchain/openai
 *   export OPENAI_API_KEY="sk-..."
 *   # Start Ferrite server: cargo run --release
 */

import { FerriteSemanticCache } from "@ferrite/ai";

async function main() {
  const cache = new FerriteSemanticCache({
    host: "127.0.0.1",
    port: 6380,
    similarityThreshold: 0.85,
    ttl: 3600,
    namespace: "lc_demo",
  });

  console.log("LangChain.js + Ferrite Semantic Cache\n");

  // Simulate LangChain cache workflow
  const llmString = "gpt-4o-mini";
  const testCases = [
    {
      prompt: "What is the capital of France?",
      expected: "Cache MISS (first call)",
    },
    {
      prompt: "France's capital city?",
      expected: "Cache HIT (semantically similar)",
    },
    {
      prompt: "Explain quantum entanglement",
      expected: "Cache MISS (different topic)",
    },
    {
      prompt: "What is quantum entanglement?",
      expected: "Cache HIT (semantically similar)",
    },
  ];

  for (const { prompt, expected } of testCases) {
    const start = performance.now();

    // Lookup
    let result = await cache.lookup(prompt, llmString);
    if (!result) {
      // Simulate LLM call
      const fakeResponse = `This is a simulated response for: ${prompt}`;
      await cache.update(prompt, llmString, [{ text: fakeResponse }]);
      result = [{ text: fakeResponse }];
    }

    const latency = performance.now() - start;

    console.log(`Query:    '${prompt}'`);
    console.log(`Response: ${result[0].text.slice(0, 80)}...`);
    console.log(`Latency:  ${latency.toFixed(1)} ms — ${expected}`);
    console.log();
  }
}

main().catch(console.error);
